src/hdmf/build/classgenerator.py

```diff
@@ -0,0 +1,327 @@
+
+    def __init__(self):
+        self.__custom_generators = []
+
+    @property
+    def custom_generators(self):
+        return self.__custom_generators
+
+    @docval({"name": "generator", "type": type, "doc": "the CustomClassGenerator class to register"})
+    def register_generator(self, **kwargs):
+        """Add a custom class generator to this ClassGenerator.
+
+        Generators added later are run first. Duplicates are moved to the top of the list.
+        """
+        generator = getargs('generator', kwargs)
+        if not issubclass(generator, CustomClassGenerator):
+            raise ValueError('Generator %s must be a subclass of CustomClassGenerator.' % generator.__name__)
+        if generator in self.__custom_generators:
+            self.__custom_generators.remove(generator)
+        self.__custom_generators.insert(0, generator)
+
+    @docval({"name": "data_type", "type": str, "doc": "the data type to create a AbstractContainer class for"},
+            {"name": "spec", "type": BaseStorageSpec, "doc": ""},
+            {"name": "parent_cls", "type": type, "doc": ""},
+            {"name": "attr_names", "type": dict, "doc": ""},
+            {"name": "type_map", "type": 'TypeMap', "doc": ""},
+            returns='the class for the given namespace and data_type', rtype=type)
+    def generate_class(self, **kwargs):
+        '''Get the container class from data type specification.
+        If no class has been associated with the ``data_type`` from ``namespace``, a class will be dynamically
+        created and returned.
+        '''
+        data_type, spec, parent_cls, attr_names, type_map = getargs('data_type', 'spec', 'parent_cls', 'attr_names',
+                                                                    'type_map', kwargs)
+
+        not_inherited_fields = dict()
+        for k, field_spec in attr_names.items():
+            if k == 'help':  # pragma: no cover
+                # (legacy) do not add field named 'help' to any part of class object
+                continue
+            if not spec.is_inherited_spec(field_spec):
+                not_inherited_fields[k] = field_spec
+        try:
+            classdict = dict()
+            bases = [parent_cls]
+            docval_args = list(deepcopy(get_docval(parent_cls.__init__)))
+            for f, field_spec in not_inherited_fields.items():
+                for class_generator in self.__custom_generators:
+                    # each generator can update classdict and docval_args
+                    if class_generator.apply_generator_to_field(field_spec, bases, type_map):
+                        class_generator.process_field_spec(classdict, parent_cls, f, field_spec, not_inherited_fields,
+                                                           type_map)
+                        class_generator.update_docval_args(docval_args, f, field_spec, type_map)
+                        break  # each field_spec should be processed by only one generator
```

Comment:
This could be combined into a single method
`class_generator.process_field_spec(...)` that returns True/False. If return True, then it has been processed, and break the loop.